File: stats.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* --- Sorting (used by median and mode) --- */
int compare_ints(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

void sort_array(int *arr, int n) {
    qsort(arr, n, sizeof(int), compare_ints);
}

/* --- Mean --- */
double calculate_mean(int *arr, int n) {
    long sum = 0;
    for (int i = 0; i < n; i++) {
        sum += arr[i];
    }
    return (double)sum / n;
}

/* --- Median --- */
double calculate_median(int *arr, int n) {
    int *sorted = malloc(n * sizeof(int));
    memcpy(sorted, arr, n * sizeof(int));
    sort_array(sorted, n);

    double median;
    if (n % 2 == 0) {
        median = (sorted[n / 2 - 1] + sorted[n / 2]) / 2.0;
    } else {
        median = sorted[n / 2];
    }

    free(sorted);
    return median;
}

/* --- Mode --- */
/* Returns all modes via out_modes (caller must free). Sets *out_count. */
int *calculate_mode(int *arr, int n, int *out_count) {
    int *sorted = malloc(n * sizeof(int));
    memcpy(sorted, arr, n * sizeof(int));
    sort_array(sorted, n);

    /* Find maximum frequency */
    int max_freq = 1, cur_freq = 1;
    for (int i = 1; i < n; i++) {
        if (sorted[i] == sorted[i - 1]) {
            cur_freq++;
            if (cur_freq > max_freq) max_freq = cur_freq;
        } else {
            cur_freq = 1;
        }
    }

    /* Collect all values that appear max_freq times */
    int *modes = malloc(n * sizeof(int));
    int count = 0;
    cur_freq = 1;
    for (int i = 1; i <= n; i++) {
        if (i < n && sorted[i] == sorted[i - 1]) {
            cur_freq++;
        } else {
            if (cur_freq == max_freq) {
                modes[count++] = sorted[i - 1];
            }
            cur_freq = 1;
        }
    }

    free(sorted);
    *out_count = count;
    return modes;
}

/* --- Main --- */
int main(void) {
    int data[] = {4, 1, 2, 2, 3, 5, 3, 3, 7, 1};
    int n = sizeof(data) / sizeof(data[0]);

    printf("Data: ");
    for (int i = 0; i < n; i++) {
        printf("%d%s", data[i], i < n - 1 ? ", " : "\n");
    }

    printf("Mean:   %.2f\n", calculate_mean(data, n));
    printf("Median: %.2f\n", calculate_median(data, n));

    int mode_count;
    int *modes = calculate_mode(data, n, &mode_count);
    printf("Mode:   ");
    for (int i = 0; i < mode_count; i++) {
        printf("%d%s", modes[i], i < mode_count - 1 ? ", " : "\n");
    }
    free(modes);

    return 0;
}

File: stats.ml
(* ---- Mean ---- *)
let calculate_mean lst =
  let sum   = List.fold_left (+) 0 lst in
  let count = List.length lst in
  float_of_int sum /. float_of_int count

(* ---- Median ---- *)
let calculate_median lst =
  let sorted = List.sort compare lst in
  let n = List.length sorted in
  if n mod 2 = 0 then
    let a = List.nth sorted (n / 2 - 1) in
    let b = List.nth sorted (n / 2) in
    (float_of_int a +. float_of_int b) /. 2.0
  else
    float_of_int (List.nth sorted (n / 2))

(* ---- Mode ---- *)
(* Build a frequency association list [(value, count); ...] *)
let build_freq lst =
  List.fold_left
    (fun acc x ->
       match List.assoc_opt x acc with
       | None   -> (x, 1) :: acc
       | Some c -> (x, c + 1) :: List.filter (fun (k, _) -> k <> x) acc)
    [] lst

let calculate_mode lst =
  let freq     = build_freq lst in
  let max_freq = List.fold_left (fun acc (_, c) -> max acc c) 0 freq in
  let modes    = List.filter_map
                   (fun (v, c) -> if c = max_freq then Some v else None)
                   freq in
  List.sort compare modes

(* ---- Helpers for printing ---- *)
let print_list sep fmt lst =
  let rec aux = function
    | []      -> ()
    | [x]     -> Printf.printf "%s" (fmt x)
    | x :: tl -> Printf.printf "%s%s" (fmt x) sep; aux tl
  in aux lst

(* ---- Main ---- *)
let () =
  let data = [4; 1; 2; 2; 3; 5; 3; 3; 7; 1] in

  print_string "Data:   ";
  print_list ", " string_of_int data;
  print_newline ();

  Printf.printf "Mean:   %.2f\n" (calculate_mean data);
  Printf.printf "Median: %.2f\n" (calculate_median data);

  print_string "Mode:   ";
  print_list ", " string_of_int (calculate_mode data);
  print_newline ()
